File: c++/LongestCommonSubsequenceOfTwo.cpp
#include "stdafx.h"
#include "LongestCommonSubsequenceOfTwo.h"

using namespace std;

int dpLCS(int n, int* A, int m, int* B)
{
	int** result = new int*[n + 1];
	for (int i = 0; i < n + 1; i++)
		result[i] = new int[m + 1];

	for (int i = 0; i < n + 1; i++) {
		for (int j = 0; j < m + 1; j++) {
			if (i == 0 || j == 0) {
				result[i][j] = 0;
			}
			else {
				int insertion = result[i][j - 1];
				int deletion = result[i - 1][j];
				int mismatchOrMatch = result[i - 1][j - 1] + (A[i - 1] == B[j - 1] ? 1 : 0);

				result[i][j] = max({ insertion, deletion, mismatchOrMatch });
			}
		}
	}

	/*for (int i = 0; i < n + 1; i++) {
		for (int j = 0; j < m + 1; j++) {
			cout << result[i][j] << " ";
		}
		cout << endl;
	}*/

	return result[n][m];
}

LongestCommonSubsequenceOfTwo::LongestCommonSubsequenceOfTwo()
{
	/*int n, m;
	cin >> n;
	int* A = new int[n];
	for (int i = 0; i < n; i++)
		cin >> A[i];

	cin >> m;
	int* B = new int[m];
	for (int i = 0; i < m; i++)
		cin >> B[i];

	cout << dpLCS(n, A, m, B) << endl;*/

	ifstream file;
	file.open("C:/Users/sohee-um/program/git_repo/Algorithmic_Design_and_Techniques/sample/5_4_lcs2.in");
	if (!file) {
		cout << "cannot open a file!" << endl;
		exit(-1);
	}

	int n, m;
	file >> n;
	int* A = new int[n];
	for (int i = 0; i < n; i++)
		file >> A[i];

	file >> m;
	int* B = new int[m];
	for (int i = 0; i < m; i++)
		file >> B[i];

	cout << dpLCS(n, A, m, B) << endl;
}


LongestCommonSubsequenceOfTwo::~LongestCommonSubsequenceOfTwo()
{
}

File: python/LongestCommonSubsequenceOfTwo.py

def dpLCS(n, A, m, B):
    result = [[0 for i in range(0, m + 1)] for j in range(0, n + 1)];

    for i in range(1, n + 1):
        for j in range(1, m + 1):
            insertion = result[i][j - 1]
            deletion = result[i - 1][j]
            mismatchOrMatch =  result[i - 1][j - 1] + 1 * (A[i - 1] == B[j - 1])
            result[i][j] = max(insertion, deletion, mismatchOrMatch)

    for r in result:
        print(r)

    return result[n][m]


n = int(input())
A = [*map(int, input().split())]
m = int(input())
B = [*map(int, input().split())]
print(dpLCS(n, A, m, B))
'''

file = open('../sample/5_4_lcs2.in')
n = int(file.readline())
A = [*map(int, file.readline().split())]
m = int(file.readline())
first.B = [*map(int, file.readline().split())]
print(dpLCS(n, A, m, first.B))
'''